storage/google/cloud/storage/blob.py

```diff
@@ -404,14 +408,22 @@ def _get_download_url(self):
         :returns: The download URL for the current blob.
         """
         if self.media_link is None:
-            download_url = _DOWNLOAD_URL_TEMPLATE.format(path=self.path)
+            base_url = _DOWNLOAD_URL_TEMPLATE.format(path=self.path)
+            scheme, netloc, path, query, frag = urlsplit(base_url)
+            query = parse_qsl(query)
+            # Only add `generation' for synthesized URLs: 'mediaLink' from
+            # server presumably already encodes it.
             if self.generation is not None:
-                download_url += u'&generation={:d}'.format(self.generation)
-            if self.user_project is not None:
-                download_url += u'&userProject={}'.format(self.user_project)
-            return download_url
+                query.append(('generation', '{:d}'.format(self.generation)))
         else:
-            return self.media_link
+            scheme, netloc, path, query, frag = urlsplit(self.media_link)
```

Comment:
It cannot:  `userProject` is not a server-settable/storable field:  it is a per-request query parameter.